thermal-control-design.md

```diff
@@ -0,0 +1,261 @@
+	speed                   = INT                            ; fan speed
+	speed_tolerance         = INT                            ; fan speed tolerance
+	speed_target            = INT                            ; fan target speed
+	led_status              = STRING                         ; fan led status
+
+### 2.3 Syslog for thermal control
+
+If there was warning raised or warning cleared, log shall be generated:
+
+	High temperature warning: PSU 1 current temperature 85C, high threshold 80C！
+	High temperature warning cleared, PSU1 temperature restore to 75C, high threshold 80C
+
+If fan broken or become up present, log shall be generated：
+
+	Fan removed warning: Fan 1 was removed from the system, potential overheat hazard!
+	Fan removed warning cleared: Fan 1 was inserted.
+
+## 3. Thermal control management
+
+Adjust cooling device according to the current temperature can be very vendor specific and some vendors already have their own implementation. In below Appendix chapter describes a Mellanox implementation. But handle the cooling device according to some predefined policies can be generic, this is part of what Thermal control management will do.
+
+This cooling device control function can be disabled if the vendor have their own implementation in the kernel or somewhere else.
+
+### 3.1 Thermal control management flow
+
+It will be a routing function to check whether the policies was hit an the fan speed need to adjust, and also run vendor specific thermal control algorithm.
+
+
+Suggest below policies that will be applied:
+
+- Set PWM to full speed if one of PS units is not present 
+
+- Set PWM to full speed if one of FAN drawers is not present or one of tachometers is broken present 
+
+- Set the fan speed to a consant value (60% of full speed) Kernel thermal control functions was disabled.
+
+FAN status led and PSU status led shall also be set accordingly when policy meet.
+
+Policy check functions will go through the device status and adjus the fan speed if necessary, these check will be preformed by calling the platform new API.
+
+Vendor specific init function and clean up function will be performed at the beginning of the task and at then end of the task.
+
+A thermal control daemon class will be deifined with above functions defined, vendors will be allowed to have their own implementation.
+
+![](https://github.com/keboliu/SONiC/blob/master/images/thermal-control.svg)
+
+### 3.2 Policy management
+
+Policies are defined in a json file for each hwsku, for example, one SKU want to apply below policies:
+
+- Thermal control algorithm control, enabled this hwsku or not, fan speed value to set if not running;
+
+- FAN absence action, suspend the algorithm or not, fan speed value to set;
+
+- PSU absence action, suspend the algorithm or not, fan speed value to set.
+
+Below is an example for the policy configuration:
+
+	{
+	    "thermal_control_algorithm": {
+		"run_at_boot_up": true,
+		"fan_speed_when_suspend": 10000
```

Comment:
For these fan speed values, maybe we should use a percentage from 0-100% instead of RPMs? Vendors can define the max fan speed in RPMs in the platform API, and then the target RPMs can be calculated based on the percentage?

Question to platform vendors: Is there any reason to prefer target speeds in RPMs instead of percentages?

Reply:
revised.

Reply:
@jleveque should we wait for some more inputs from other vendors or we can move forward?

Reply:
@padmanarayana: Have you had time to review this PR? Please let us know if you do not have any comments. Thanks!

Reply:
@padmanarayana any comments on this design?